File: frontend/src/service/ticketService.ts
import axios from "axios";
import { Ticket } from "../dataModels/Ticket";
import { devEnvironment } from "../environments/devEnvironments";
import { ListaTickets } from "../dataModels/ListaTickets";

export interface CreateTicketDTO {
    titulo: string,
    descricao: string,
    prioridade: number,
    produtoId: number
}

export interface UpdateTicketDTO {
    titulo: string,
    descricao: string,
    prioridade: number,
    produtoId: number
}

const TICKETS_URL = devEnvironment.BACKEND_URL + "/Tickets";

export class TicketService {

    async getTickets(abortController?: AbortController) :  Promise<ListaTickets>{
        let res = await axios.request<ListaTickets>({
            method: 'GET',
            url: TICKETS_URL,
            signal: abortController?.signal
        })
        
        return res.data;
    
    }

    async getTicketById(id: number, abortController?: AbortController) :  Promise<Ticket>{
        let res = await axios.request<Ticket>({
            method: 'GET',
            url: TICKETS_URL+ `/${id}`,
            signal: abortController?.signal
        })
        
        return res.data;
    }
    
    async createTicket(ticket: CreateTicketDTO){
        
        let res = await axios.request<Ticket>({
            method: 'POST',
            url: TICKETS_URL,
            data: ticket
        })
        return res;
    }

    async updateTicket(id: number, newTicket : UpdateTicketDTO){

        let res = await axios.request<Ticket>({
            method: 'PUT',
            url: TICKETS_URL + `/${id}`,
            data: newTicket
        })

        return res;
    }
}

File: backend/Services/ImplServices/TicketService.cs
using backend.DTO;
using backend.DTO.Ticket;
using backend.Exceptions;
using backend.Execptions;
using backend.Mapper;
using backend.Models;
using backend.Repos.IRepos;
using backend.Services.IServices;

namespace backend.Services.ImplServices;

public class TicketService : ITicketService
{
    private ITicketRepo ticketRepo;
    private IProductRepo productRepo;

    public TicketService(ITicketRepo ticketRepo, IProductRepo productRepo)
    {
        this.ticketRepo = ticketRepo;
        this.productRepo = productRepo;
    }

    public async Task<TicketDTO> CreateTicket(CreateTicketDTO newTicket)
    {

        var product = await productRepo.getProductById(newTicket.ProdutoId);

        if(product == null) {
            throw new ProductNotFoundException("Product not found");
        }
        //TODO prly should move this for now it stays here
        if (newTicket.Prioridade < 1 || newTicket.Prioridade > 5) {
            throw new InvalidPriorityException("Priority should be between 1 and 5");
        }

        var ticket = await ticketRepo.createTicket(new Ticket {
            Titulo = newTicket.Titulo,
            Descricao = newTicket.Descricao,
            Prioridade = newTicket.Prioridade,
            ProdutoId = newTicket.ProdutoId,
        });

        return TicketMapper.ToTicketDTO(ticket);
    }


    public async Task<RepositoryResponse<TicketDTO>> GetTickets(QueryOptionsDTO queryOptionsDTO)
    {

        var repoRequest = new RepositoryRequest {
            Page = queryOptionsDTO.Page,
            PageLength = queryOptionsDTO.PageLength
        };

        var repoRes =  await ticketRepo.getTickets(repoRequest);

        return new RepositoryResponse<TicketDTO> {
            Items = repoRes.Items.Select(TicketMapper.ToTicketDTO), // Transforms each item into a TicketDTO
            TotalItems = repoRes.TotalItems
        }; 
        
    }
    public async Task<TicketDTO?> GetTicketById(int id)
    {
        var ticket = await ticketRepo.getTicketById(id);

        if(ticket == null) {
            throw new TicketNotFoundException("Ticket not found");
        }

        return TicketMapper.ToTicketDTO(ticket);

    }

    public async Task<TicketDTO?> UpdateTicket(int id, UpdateTicketDTO newticket)
    {
        //FInd
        var ticket = await ticketRepo.getTicketById(id);

        if(ticket == null) {
            throw new TicketNotFoundException("Ticket not found");
        }

        //Update
        ticket.Titulo = newticket.Titulo;
        ticket.Descricao = newticket.Descricao;
        ticket.Prioridade = newticket.Prioridade;
        ticket.ProdutoId = newticket.ProdutoId;

        //Save
        var updatedTicket = await ticketRepo.updateTicket(ticket);

        if(updatedTicket == null) {
            throw new Exception("Ticket update failed");
        }
    
        return TicketMapper.ToTicketDTO(updatedTicket);
    }
}